wireless_modules/middle_module/mpu.py

```diff
@@ -0,0 +1,95 @@
+        the sensor `n_samples` times and returning the mean.
+
+        If passed a `calibration` dictionary, subtract these
+        values from the final sensor value before returning.
+        :param n_samples: The number of samples to take average from
+        :param calibration: calibration values to offset by
+        :return: A dictionary of mean sensor measurements for the temperature, 3 dimensions of acceleration and
+                gyroscope
+        """
+        result = {}
+        for _ in range(n_samples):
+            data = self.accelerometer.get_values()
+
+            for key in data.keys():
+                # Add on value / n_samples to produce an average
+                # over n_samples, with default of 0 for first loop.
+                result[key] = result.get(key, 0) + (data[key] / n_samples)
+
+        if calibration:
+            # Remove calibration adjustment.
+            for key in calibration.keys():
+                result[key] -= calibration[key]
+
+        return result
+
+    def calibrate(self, threshold=50, n_samples=10):
+        """
+        code from: https://www.twobitarcade.net/article/3-axis-gyro-micropython/
+        Get calibration data for the sensor, by repeatedly measuring while the sensor is stable. The resulting
+        calibration dictionary contains offsets for this sensor in its current position.
+        :param threshold: The accuracy of the calibration
+        :param n_samples: The number of times the sensor should be read and averaged
+        """
+        print("--------Calibrating:")
+        while True:
+            v1 = self.get_smoothed_values(n_samples)
+            v2 = self.get_smoothed_values(n_samples)
+
+            # Check all consecutive measurements are within the threshold. We use abs() so all calculated
+            # differences are positive.
+            if all(abs(v1[key] - v2[key]) < threshold for key in v1.keys()):
+                self.calibrated_values = v1
+                return v1  # Calibrated.
+            print("in calibration, keep device at rest...")
+
+    def read(self):
+        """
+        Read averaged and calibrated sensor data for the accelerometer and gyroscope
+        :return: An array of length 2 containing a dictionary of acceleration values and a dictionary for
+                gyroscope values. Each contains a `type` key associated with a string of the measurement type and a
+                `value` key associated with another dictionary containing (key, value) pair of the axis and it's
+                relevant data.
+        """
+        all_data = self.get_smoothed_values(n_samples=self.samples, calibration=self.calibrated_values)
+        accel_values = {
+            "x": all_data["AcX"],
+            "y": all_data["AcY"],
+            "z": all_data["AcZ"]
+        }
+        gyro_values = {
+            "x": all_data["GyX"],
+            "y": all_data["GyY"],
+            "z": all_data["GyZ"]
+        }
```

Comment:
We should convert these into some normal units! @khanguslee what did we use on V2? Not sure if we should use Gs or m/s^2, and rpm vs deg/s vs rad/s.

For now, we can convert to Gs by dividing by 16384 and to deg/s by dividing by 131 [source](https://www.twobitarcade.net/article/3-axis-gyro-micropython/)

Reply:
Cool! I believe by default, the MPU-6050 has the acceleration range set to 2G, and gyro set to 250. I already messaged @MatthewMadsen earlier today asking what range would be required for measuring crash loads, but interesting to see we went for the highest possible for V2. Greater ranges apparently are less accurate, so we'd want to keep the range as low as is still useful.

Also, I have realised with a 1 Hz sample rate, we're probably not going to get much data from a crash... something to think about

Reply:
Do all sensors need to be queried at the same rate, otherwise we could query mpu6050 more often so we can gather better crash data. And if we don't want to affect other stuff by querying the mpu6050 more often, we could make the wireless module publish on a seperate topic for all mpu6050 data along with publishing on the existing data topic at 1Hz with all sensor values.

Reply:
Yeah I reckon it's likely we'll want to do something like that in the future. For the scope of this PR though, let's just query it at the same rate as other sensors. Something for another PR would be increase the accelerometer and gyroscope's ranges, because I don't think the library supports that as is.